source/extensions/filters/common/rbac/engine.h

```diff
@@ -45,6 +47,35 @@ class RoleBasedAccessControlEngine {
    */
   virtual bool allowed(const Network::Connection& connection, const StreamInfo::StreamInfo& info,
                        std::string* effective_policy_id) const PURE;
+
+  /**
+   * Returns whether or not the current action should be logged.
+   *
+   * @param connection the downstream connection used to identify the action/principal.
+   * @param headers    the headers of the incoming request used to identify the action/principal. An
+   *                   empty map should be used if there are no headers available.
+   * @param info       the per-request or per-connection stream info with additional information
+   *                   about the action/principal.
+   * @param effective_policy_id  it will be filled by the matching policy's ID,
+   *                   which is used to identity the source of the allow/deny.
+   */
+  virtual LogDecision shouldLog(const Network::Connection& connection,
+                                const Envoy::Http::RequestHeaderMap& headers,
+                                const StreamInfo::StreamInfo& info,
+                                std::string* effective_policy_id) const PURE;
+
+  /**
+   * Returns whether or not the current action should be logged.
+   *
+   * @param connection the downstream connection used to identify the action/principal.
+   * @param info       the per-request or per-connection stream info with additional information
+   *                   about the action/principal.
+   * @param effective_policy_id  it will be filled by the matching policy's ID,
+   *                   which is used to identity the source of the allow/deny.
+   */
+  virtual LogDecision shouldLog(const Network::Connection& connection,
+                                const StreamInfo::StreamInfo& info,
+                                std::string* effective_policy_id) const PURE;
```

Comment:
This seems pretty redundant with above. Can we modify above to have a single set of functions called something like `evaluate` which return an enum around allow/deny/log?

Reply:
The way I have it, all actions call both functions, and each function has different logic depending on the action, since currently the filter doesn't check the action and leaves that to the engine. This isn't ideal though, since you're right there is alot of redundancy. I'll try to restructure this somehow.